shiny/_app.py

```diff
@@ -446,3 +452,29 @@ def file_response_handler(req: Request) -> FileResponse:
             file_response_handler,
             name="shiny-app-static-assets-" + mount_point,
         )
+
+
+def noop_server(input: Inputs, output: Outputs, session: Session) -> None:
+    pass
+
+
+def ignore_output_session_server(
+    server: Callable[[Inputs], None]
+) -> Callable[[Inputs, Outputs, Session], None]:
+    def _server(input: Inputs, output: Outputs, session: Session):
+        # Only has 1 parameter, ignore output, session
+        server(input)
+
+    return _server
+
+
+def has_only_input(
+    x: Callable[[Inputs], None] | Callable[[Inputs, Outputs, Session], None]
+) -> TypeGuard[Callable[[Inputs], None]]:
+    return len(signature(x).parameters) == 1
```

Comment:
I understand that the `TypeGuard` does provide some help for the type checker here, but still, I think it's clearer if the `len(...) == 1` check is inlined in the `elif` statements above, and then a `cast` is used inside the `elif` blocks. As it is, I find myself jumping up and down in the code to understand what it's doing.

It would be nice if Python type checking were smarter and didn't require either of these workarounds.